### ISSUE TITLE
UI.Next client-server example throws an exception

### ISSUE THREAD

**Issue by Houndolon** [intellifactory/websharper.ui.next#89]
When the example at the bottom of page [https://github.com/intellifactory/websharper.ui.next/blob/master/docs/ClientServer.md](https://github.com/intellifactory/websharper.ui.next/blob/master/docs/ClientServer.md) with a slightly changed `Main` function is run in Zafir, it throws an exception.

The `Main` function is:

```
[<Website>]
let Main =
    let pageBody = [
        h1 [text "Who are you?"]
        div [ client <@ ClientCode.InputControl() @> ]
        inputAttr [
            attr.``type`` "submit"
            on.click <@ ClientCode.Submit @>
        ] []
        div [ client <@ ClientCode.OutputControl() @> ]
    ]

    let createContentPage context = 
        Content.Page(
            Title = "UI.Next client-server example",
            Body = pageBody
        )

    Application.SinglePage createContentPage
```

The stack trace is:

```
[Exception: Invalid handler function: Lambda (el, Lambda (ev, Call (None, Submit, [el, Coerce (ev, Event)])))]
   <StartupCode$WebSharper-UI-Next>.Handler@101-5.Invoke(String message) +33
   WebSharper.UI.Next.Attr.Handler(String event, FSharpExpr`1 callback) +759
   WebSharper.UI.Next.on.click(FSharpExpr`1 f) +18
   <StartupCode$Application3>.$Main..cctor() in D:\Solutions\UINextApplication1\Application3\Main.fs:39

[TypeInitializationException: The type initializer for '<StartupCode$Application3>.$Main' threw an exception.]
   Application3.ServerCode..cctor() in D:\Solutions\UINextApplication1\Application3\Main.fs:0

[TypeInitializationException: The type initializer for 'Application3.ServerCode' threw an exception.]
   Application3.ServerCode.get_Main() +0

[TargetInvocationException: Exception has been thrown by the target of an invocation.]
   System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor) +0
   System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments) +87
   System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) +101
   WebSharper.Sitelets.TryLoadSiteB@60-1.Invoke(PropertyInfo p) +112

[Exception: Failed to initialize sitelet definition: Application3.ServerCode.Main]
   WebSharper.Sitelets.TryLoadSiteB@60-1.Invoke(PropertyInfo p) +465
   Microsoft.FSharp.Collections.ArrayModule.loop@263-30(FSharpFunc`2 f, T[] array, Int32 i) +45
   Microsoft.FSharp.Collections.ArrayModule.TryPick(FSharpFunc`2 chooser, T[] array) +59
   Microsoft.FSharp.Collections.SeqModule.TryPick(FSharpFunc`2 chooser, IEnumerable`1 source) +142
   WebSharper.Sitelets.SiteLoading.TryLoadSite(Assembly assembly) +33
   WebSharper.Sitelets.LoadFromAssemblies@89-1.Invoke(Assembly assembly) +7
   Microsoft.FSharp.Collections.choose@154.System-Collections-IEnumerator-MoveNext() +113
   Microsoft.FSharp.Collections.SeqModule.ToList(IEnumerable`1 source) +223
   WebSharper.Sitelets.LoadFromAssemblies@84.Invoke(Unit unitVar0) +72
   WebSharper.Sitelets.Extensions.Timed(a message, FSharpFunc`2 action) +105
   WebSharper.Sitelets.SiteLoading.LoadFromAssemblies(HttpApplication app) +28
   WebSharper.Sitelets.HttpModule.System-Web-IHttpModule-Init(HttpApplication app) +36
   System.Web.HttpApplication.RegisterEventSubscriptionsWithIIS(IntPtr appContext, HttpContext context, MethodInfo[] handlers) +536
   System.Web.HttpApplication.InitSpecial(HttpApplicationState state, MethodInfo[] handlers, IntPtr appContext, HttpContext context) +172
   System.Web.HttpApplicationFactory.GetSpecialApplicationInstance(IntPtr appContext, HttpContext context) +336
   System.Web.Hosting.PipelineRuntime.InitializeApplication(IntPtr appContext) +296

[HttpException (0x80004005): Failed to initialize sitelet definition: Application3.ServerCode.Main]
   System.Web.HttpRuntime.FirstRequestInit(HttpContext context) +9963380
   System.Web.HttpRuntime.EnsureFirstRequestInit(HttpContext context) +101
   System.Web.HttpRuntime.ProcessRequestNotificationPrivate(IIS7WorkerRequest wr, HttpContext context) +254
```

Curiously enough, if the line

```
on.click <@ ClientCode.Submit @>
```

is replaced with

```
Attr.Handler "click" <@ ClientCode.Submit @>
```

it works as expected.

Also, changing the `ClientCode.Submit` function from:

```
    let Submit (el: Dom.Element) (ev: Dom.Event) =
        submit.Trigger()
```

to:

```
    let Submit (el: Dom.Element) (ev: Dom.MouseEvent) =
        submit.Trigger()
```

makes the original `on.click` version work.

I don't know if this is the intended behaviour and the error is in the example code, or `on.click` is misbehaving. I also have no idea how the original version manages to compile since `on.click` expects a `JavaScript.Dom.Element -> JavaScript.Dom.MouseEvent -> unit`, but gets a `JavaScript.Dom.Element -> JavaScript.Dom.Event-> unit` which seems to be a generalized version of what `on.click` expects.

**Comment by Tarmil** [comment]
Thanks for the report! This should be a quick fix, we simply need to ignore the implicit upcast that takes place when passing a `MouseEvent` to a function that expects an `Event`.

**Comment by Tarmil** [comment]
This fix needs to be backported to WebSharper 3.6.
